### ISSUE TITLE
Provide an API for Awaitable Combinators

### ISSUE THREAD

**Issue by martinschroeder** [concurrent-php/task#14]
The `Deferred` class should support creation of awaitable combinators through a simple API that is safe to use as far as error handling is concerned. See #2 for an explanation of the basic workings. The proposed API looks like this:

```php
final class Deferred
{
```

```php
public static function each(callable(Deferred, int, ?\Throwable, mixed = null) $continuation, Awaitable ...$awaitables): Awaitable { }
```
This would basically specify a map function to be applied to the given awaitables. The function would be registered as continuation and called once for every awaitable as it succeeds or fails. The first argument would be a `Deferred` that is used to succeed or fail the `Awaitable` returned from the helper method. The second argument would be the numeric index of the input awaitable. Arguments 3 and 4 are error and success value emitted by the current awaitable.

If an error is thrown from the callback function it will automatically fail the `Deferred` (and hence the result `Awaitable`, that way no contextual error handling is needed. If all awaitables have run to completion and the `Deferred` has not been resolved it will automatically fail.

The method could also accept an `array` or `iterable` as second argument instead of using varargs which would change the index parameter type to mixed. Eighter way the `continueWith()` method would be gone and `Awaitable` would just be a marker interface.

@kelunik What do you think about this? It believe it addresses your concerns about error handling (which could be dropped from `Context`) and allows for arbitrary combinators and other stuff to be implemented in a safe way.

**Comment by kelunik** [comment]
I'm not entirely sure whether we need that. The amount of combinators that will exist will be rather small. I guess these could all be implemented in core without the need for any custom combinators.

**Comment by martinschroeder** [comment]
It could be useful to provide a specialized single-awaitable form, that is basically just a transformation working on the outcome of an input awaitable. This could be implemented like so:
```php
namespace Concurrent;

final class Deferred
{
    public static function transform(Awaitable $awaitable, callable $continuation): Awaitable { }
}
```
In this case the `continuation` callback would have a signature of `callable(?\Throwable, mixed = null)`. It would just receive the outcome of the input awaitable and eighter return a value or throw an error.

**Comment by kelunik** [comment]
We killed that in Amp. Sometimes I miss it, because we have to use a new generator, but it's not really a big real doing it with `async` / `await`.

**Comment by trowski** [comment]
I agree with @kelunik that a combinator API seems unnecessary. `Deferred` plus `async` / `await` is sufficient to implement any combinator.

**Comment by martinschroeder** [comment]
While `async` / `await` is sufficient to implement combinator in userland it is not an efficient solution. Let me use an `all()` combinator function as example to outline why i am implementing a different solution to that problem. I left out input validation out to make the example a little shorter and focus on the async aspect.

```php
function all(array $awaitables): Awaitable
{
    $defer = new Deferred();
    
    $result = \array_fill_keys(\array_keys($awaitables), null);
    $remaining = \count($awaitables);
    
    foreach ($awaitables as $k => $a) {
        Task::async(function () use ($k, $a, $defer, & $result, & $remaining) {
            try {
                $result[$k] = Task::await($a);
                
                if (--$remaining === 0) {
                    $defer->resolve($result);
                }
            } catch (\Throwable $e) {
                $defer->fail($e);
            }
        });
    }
    
    return $defer->awaitable();
}
```
This is a pretty much standard implementation of such a combinator. Since the public API of `Awaitable` has been removed in #8 there is no way to "subscripe" to the input awaitables. This requires the use of an additional `Task` for each awaitable to access the result. Therefore additional resources have to be allocated for each task object (this includes allocating a new C & PHP stack!). The scheduler has to perform a context switch and execute the new callback until it reaches the call to `await()`.

If the awaitable has already completed execution it will just copy the result and resume execution. In both these cases the task is basically created and destroyed in one go just to access the result value. In case of running tasks the new task needs to suspend itself until the awaited task completes, which is not useful eighter because it uses up resources without actively doing anything. In case of tasks that are not started yet it is a little more efficient due to task-inlining (in this case only resource allocation for the task itself is needed, no new stack will be allocated).

```php
function all(array $awaitables): Awaitable
{
    $result = \array_fill_keys(\array_keys($awaitables), null);
    
    $all = function (Deferred $defer, $last, $k, $e, $v) use (& $result) {
        if ($e) {
            $defer->fail($e);
        } else {
            $result[$k] = $v;
            
            if ($last) {
                $defer->resolve($result);
            }
        }
    };
    
    return Deferred::combine($awaitables, $all);
}
```

The second example shows a much more efficient way to achieve the same thing. Internally not a single `Task` will be needed. The only overhead involved is creating the `Deferred` object (which is also needed in the first example) and storing the callback function along with the input awaitable count. The continuation mechanism being used relies on a C function pointer (kind of a callback) in case of running tasks and even that is and that is a lot faster and more memory-efficient.
